Add Product Image Gallery

gigitux opened the pull request:
This PR adds the Product Image Gallery block. It is possible to add this block only on the Single Product Image Gallery thanks to [this](https://github.com/woocommerce/woocommerce-blocks/blob/1d63441d7a14bf1dd4eabf2c83f2480e965ca467/assets/js/atomic/utils/register-block-single-product-template.ts).


| Block Details | Preview Editor Side | Frontend  |
|---------------|---------------------|-----------|
|![image](https://user-images.githubusercontent.com/4463174/215522884-09459aa4-e51b-4975-9a52-68e4a5a3f769.png)|![image](https://user-images.githubusercontent.com/4463174/218435020-62a15946-2932-40d9-86ec-309bde156c2d.png)|![image](https://user-images.githubusercontent.com/4463174/215523026-a0a35211-ba40-49db-bb88-c773826aa7a1.png)|

### Design Feedback
@vivialice could give me feedback about:
1. Icon to use.
2. Description.
3. Preview on the editor side.


### Notes for testing this PR.

Because of #8314, the styles of this block are not loaded. To avoid this issue, add the class `product` to the main div that wraps the entire page.

Fixes #8233

### Testing


#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Via Site Editor, edit the Single Product Template.
2. Add the `Product Image Gallery`. Save it.
3. On the frontend side, open a product page, check that the Product Image Gallery is loaded, and show the right images.
4. Check that the block can be added only on the Single Product Template

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Add Product Image Gallery block.

gigitux committed:
Add Product Image Gallery #8233

Add Product Image Gallery

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

gigitux committed:
Add Product Image Gallery block

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

gigitux committed:
remove support global styles

gigitux committed:
remove support global styles

gigitux committed:
Merge branch 'add/8233-product-image-gallery-block' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

gigitux committed:
Merge branch 'trunk' into add/8233-product-image-gallery-block

imanish003 commented on assets/js/atomic/blocks/product-elements/product-image-gallery/edit.tsx:
```diff
@@ -0,0 +1,72 @@
+				style={ {
+					width: '500px',
+					height: '500px',
+				} }
```
Is there any specific reason we are adding width & height of `img` here as well as inside `editor.scss` ?

gigitux replied:
I forgot to remove the inline style :| 
I did the fix via [795a698](https://github.com/woocommerce/woocommerce-blocks/pull/8235/commits/795a6980d6e084b183b7499f163689df18e4bad2) commit.

imanish003 commented on assets/js/atomic/blocks/product-elements/product-image-gallery/editor.scss:
```diff
@@ -0,0 +1,12 @@
+		width: 500px;
+		height: 500px;
+	}
+	.wc-block-editor-product-gallery__gallery {
```
I am nitpicking here but maybe we can improve the naming of this class. Some of alternatives could be:
- `wc-block-editor-product-gallery__alternative-images`
- `wc-block-editor-product-gallery__other-images`

If it doesn't break the frontend.

gigitux replied:
Thanks for the suggestion! I picked `wc-block-editor-product-gallery__other-images`. [795a698](https://github.com/woocommerce/woocommerce-blocks/pull/8235/commits/795a6980d6e084b183b7499f163689df18e4bad2)

imanish003 commented on assets/js/atomic/utils/register-block-single-product-template.ts:
```diff
@@ -0,0 +1,46 @@
+} ) => {
+	let currentTemplateId: string | undefined;
+
+	subscribe( () => {
```
Maybe we can add 2nd argument to subscribe function i.e. storeNameOrDescriptor parameter, the listener function will be called only on updates on that one specific registered store. In our case, I believe it would be `core/edit-site`. 
What do you think?

gigitux replied:
Maybe I don't understand what do you mean, but the `subscribe` function receive only the callback as parameter.

imanish003 replied:
Sorry for not being clear.

I checked [this doc](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#subscribe) and it looks like there is 2nd argument that could be passed. Here is the text from the link:

```
Given a listener function, the function will be called any time the state value
of one of the registered stores has changed. If you specify the optional
storeNameOrDescriptor parameter, the listener function will be called only
on updates on that one specific registered store.

This function returns an unsubscribe function used to stop the subscription.

Usage

import { subscribe } from '@wordpress/data';

const unsubscribe = subscribe( () => {
    // You could use this opportunity to test whether the derived result of a
    // selector has subsequently changed as the result of a state update.
} );

// Later, if necessary...
unsubscribe();

Parameters

    listener Function: Callback function.
    storeNameOrDescriptor string|StoreDescriptor?: Optional store name.
```

Please let me know if I am missing something.

gigitux replied:
Oh, nice! I updated the PR and also I created a PR on DefinitelyTyped repo https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64177

imanish003 replied:
Thanks for creating a PR on DefinitelyTyped repo 🙌🏻

imanish003 commented on src/BlockTypes/ProductImageGallery.php:
```diff
@@ -0,0 +1,63 @@
+	protected function render( $attributes, $content, $block ) {
+
+		$post_id = $block->context['postId'];
+		global $product;
```
Is there any specific reason behind defining `$product` as global here?

gigitux replied:
Yep. It is necessary, otherwise, `woocommerce_show_product_images()` doesn't work.

imanish003 replied:
Got it. Thanks for explaining it to me 🤝

imanish003 reviewed (approved):
Hi @gigitux, I have tested the PR & everything is working as expected after adding `product` class. Excellent work 🙌🏻 💪🏻

I have left a few minor suggestions, but those shouldn't block this PR. Therefore, I am pre-approving the PR.

gigitux committed:
address CSS feedback

gigitux committed:
Merge branch 'add/8233-product-image-gallery-block' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

gigitux committed:
add support for the custom classname

gigitux commented:
Thanks for the review and feedback! I addressed your comments.
Also, I added the custom class support via bfd2e98f615e894ce84fb3dde1d6d2b6862ac603 commit.

gigitux commented:
Before merging the PR, I prefer to wait for feedback from @vivialice.

gigitux committed:
remove save function

gigitux commented:
@imanish003 JFYI I removed the save function because it is not necessary for the server-side blocks 👍 [e526eff](https://github.com/woocommerce/woocommerce-blocks/pull/8235/commits/e526eff91a4bd6e955e62793f5b18c259305e5ce)

dinhtungdu commented on assets/js/atomic/utils/register-block-single-product-template.ts:
```diff
@@ -0,0 +1,46 @@
+} ) => {
+	let currentTemplateId: string | undefined;
+
+	subscribe( () => {
```
Could we use debounce here to make this trigger less frequent?

gigitux replied:
Mmm, I would avoid a debounce here to avoid any kind of race condition.

dinhtungdu replied:
I'm curious what can be the race condition and its drawback. The motivation behind my suggestion is to keep the editor performant.

gigitux replied:
Mmm, for example, the user is editing another template, but the function isn't called due to the debounce and the block is still visible in the inserter.

dinhtungdu replied:
> but the function isn't called due to the debounce

I don't think we will have this issue, as we just want to debounce them (by 100ms, for example), the subscription still triggers, but less frequently. So when the user switches to another template, the subscription may not trigger right away, but still fast enough to register/unregister the block before user interaction.

We don't have to do it now btw, we can go with the current PR and come back to this later when we see it affect the editor performance.

gigitux committed:
add second parameter to the subscribe function

vivialice commented:
Hey @gigitux ! 👋 

For the **block icon**, I think it's best to use the same `gallery` block as core. It is in line with other feedback we have received, so that works best until we do more with this block.

I took a look at **the preview**, and we should align with the greys used in wireframe layouts in core. However, I did notice there are two styles being used to represent images: 

_Screenshot of one of the Wireframe pattern image placeholders:_
<img width="620" alt="Screenshot 2023-02-02 at 14 09 56" src="https://user-images.githubusercontent.com/9888201/216347269-17fe7ccd-da1f-436f-9533-a1cf230aeb2d.png">

_Screenshot of the Image block placeholder:_ 
<img width="653" alt="Screenshot 2023-02-02 at 14 10 31" src="https://user-images.githubusercontent.com/9888201/216347407-1a8fcb52-ea7a-42f4-843d-3fd4ab4834fb.png">

I think it's best to align with the Wireframe style. I have created a mockup: 

<img width="712" alt="Screenshot 2023-02-02 at 14 12 26" src="https://user-images.githubusercontent.com/9888201/216347849-a9d4a1b7-3cb9-4260-9b58-92465732c838.png">

Do you need an image or svg to implement the above? 

For the **block description**, let's use `Display a product's images.`

gigitux commented:
Thanks @vivialice for the feedback! 🙇 

>Do you need an image or svg to implement the above?

Could you share the SVG?

vivialice commented:
Sure thing! Here you go @gigitux 

[Download](https://cloudup.com/coNTr6RFUdh)

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

gigitux commented:
> Sure thing! Here you go @gigitux
> 
> [Download](https://cloudup.com/coNTr6RFUdh)

Ops! Sorry, I didn't specify. Could you share only this piece of the image?
![image](https://user-images.githubusercontent.com/4463174/216689514-cc5bbc54-ec1b-43e2-bfaf-0a158849c305.png)

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

gigitux committed:
update @types/wordpress__data package

gigitux commented:
Via 8aee84125143d2ecf3eaf3af1b74ca449b034cb2 I updated the package `@types/wordpress__data` and removed the comment that I added after [this discussion](https://github.com/woocommerce/woocommerce-blocks/pull/8235#discussion_r1091730765).

gigitux commented:
@vivialice I updated the image, this is the preview:

![image](https://user-images.githubusercontent.com/4463174/218434690-9059220d-5ad0-4836-a094-d6f094852a4e.png)

I'm going to merge the PR, feel free to comment here so that I can create a follow-up PR addressing your feedback! 🙇

gigitux committed:
update placeholder, icon and description

gigitux committed:
update tsconfig

gigitux committed:
Merge branch 'trunk' into add/8233-product-image-gallery-block
